app/src/lib/stores/alive-store.ts

```diff
@@ -0,0 +1,217 @@
+    // Clear subscriptions for accounts that are no longer in the list
+    for (const account of subscribedAccounts) {
+      if (!accountIncluded(account, accounts)) {
+        this.unsubscribeFromAccount(account)
+      }
+    }
+
+    // Subscribe to new accounts
+    for (const account of accounts) {
+      if (!accountIncluded(account, subscribedAccounts)) {
+        this.subscribeToAccount(account)
+      }
+    }
+  }
+
+  private sessionForAccount(
+    account: Account
+  ): IAliveEndpointSession | undefined {
+    return this.sessionPerEndpoint.get(account.endpoint)
+  }
+
+  private async createSessionForAccount(
+    account: Account
+  ): Promise<IAliveEndpointSession | null> {
+    const session = this.sessionForAccount(account)
+    if (session !== undefined) {
+      return session
+    }
+
+    const api = API.fromAccount(account)
+    const webSocketUrl = await api.getAliveWebSocketURL()
+
+    if (webSocketUrl === null) {
+      return null
+    }
+
+    const aliveSession = new AliveSession(
+      webSocketUrl,
+      api.getAliveWebSocketURL,
+      false,
+      this.notify
+    )
+
+    const newSession = {
+      session: aliveSession,
+      webSocketUrl,
+    }
+
+    this.sessionPerEndpoint.set(account.endpoint, newSession)
+
+    return newSession
+  }
+
+  private unsubscribeFromAccount(account: Account) {
+    const endpointSession = this.sessionForAccount(account)
+    if (endpointSession === undefined) {
+      return
+    }
+
+    const subscription = this.subscriptions.find(s =>
+      accountEquals(s.account, account)
+    )
+    if (subscription === undefined) {
+      return
+    }
+
+    endpointSession.session.unsubscribe([subscription.subscription])
+    this.subscriptions = this.subscriptions.filter(
+      s => !accountEquals(s.account, account)
+    )
+
+    this.sessionPerEndpoint.delete(account.endpoint)
+
+    endpointSession.session.offline()
+
+    console.log('Unubscribed from Alive channel!')
+  }
+
+  private subscribeToAccount = async (account: Account) => {
+    const endpointSession = await this.createSessionForAccount(account)
+    const api = API.fromAccount(account)
+    const channelInfo = await api.getAliveDesktopChannel()
+
+    if (endpointSession === null || channelInfo === null) {
+      return
+    }
+
+    const subscription = {
+      subscriber: this,
+      topic: {
+        name: channelInfo.channel_name,
+        signed: channelInfo.signed_channel,
+        offset: '', // TODO: do we need to set an offset?
+      },
+    }
+
+    endpointSession.session.subscribe([subscription])
+
+    this.subscriptions.push({
+      account,
+      subscription,
+    })
+
+    console.log('Subscribed to Alive channel!')
+  }
+
+  private notify = (subscribers: Iterable<AliveStore>, event: AliveEvent) => {
+    console.log('Alive event received:', event)
+
+    if (event.type !== 'message') {
+      return
+    }
+
+    // TODO: parse this safely
+    const desktopEvent = event.data as any
```

Comment:
@niik is there a proper/cleaner way of handling this in TypeScript?

Technically, we could receive anything in that `data`. According to the type definitions, it should actually conform to a specific interface with these fields:

```typescript
    timestamp: number;
    wait: number;
    gid?: string;
```

But I don't see any of them being used in the `@github/alive-client` package, so I don't think it's required for us (I have to double check).

Thoughts?

Reply:
There are sadly no ways (that I know of) of taking a runtime object and checking its conformity to a compile-time type. At some point you need to tell typescript that the object can be anything and then decide how much you trust the service on the other end. You can certainly type check (at runtime) all the properties to build up a new object which is guaranteed to conform. Or you can toss all safety to the wayside and say "trust me" which is essentially what you're doing here.

So, in decreasing order of safety:

## Testing everything

```ts
const data = event.data as any
    if (data.type === ('pr-checks-failed' as const)) {
      const {
        timestamp,
        owner,
        repo,
        pull_request_number,
        check_suite_id,
        commit_sha,
      } = data

      assert(typeof timestamp === 'number', 'Expected timestamp to be numeric')
      assert(typeof owner === 'string', 'Expected owner to be a string')
      assert(typeof repo === 'string', 'Expected repo to be a string')
      assert(typeof pull_request_number === 'number', 'Expected pull_request_number to be a string')
      assert(typeof check_suite_id === 'number', 'Expected check_suite_id to be a string')
      assert(typeof commit_sha === 'string', 'Expected commit_sha to be a string')

      const desktopEvent: IDesktopChecksFailedAliveEvent = {
        type: 'pr-checks-failed',
        timestamp,
        owner,
        repo,
        pull_request_number,
        check_suite_id,
        commit_sha,
      }

      this.emitter.emit(this.ALIVE_EVENT_RECEIVED_EVENT, desktopEvent)
    }
```

## Test that the type matches what we expect it to and then trust the rest

```
    const data = (event.data as any) as DesktopAliveEvent
    if (data.type === 'pr-checks-failed') {
      this.emitter.emit(this.ALIVE_EVENT_RECEIVED_EVENT, data)
    }
```

## YOOOOLOOOO

```
this.emitter.emit(this.ALIVE_EVENT_RECEIVED_EVENT, event.data as any)
```


My vote would be on number 2 here.